tests/x11/nis_server.pm

```diff
@@ -126,15 +125,18 @@ sub run {
     nis_server_configuration();
     wait_serial("$module_name-0", 360) || die "'yast2 nis server' didn't finish";
     assert_screen 'yast2_closed_xterm_visible';
-    # NIS Server is configured and running, configuration continues on client side
-    mutex_create('nis_ready');
     $module_name = y2_module_consoletest::yast2_console_exec(yast2_module => 'nfs_server');
     nfs_server_configuration();
     wait_serial("$module_name-0", 360) || die "'yast2 nfs server' didn't finish";
     assert_screen 'yast2_closed_xterm_visible', 200;
-    # NFS Server is configured and running, configuration continues on client side
-    mutex_create('nfs_ready');
-    wait_for_children;
+    # In order for the hostname to get the set value via yast2 nis_server, a restart is needed.
+    # Otherwise "make" command won't work as in Makefile, there is a variable that gets it's value from "domainname" command
+    systemctl 'restart network';
+    # NIS and NFS Server is configured and running, configuration continues on client side
+    mutex_create('server_ready');
+    my $children = get_children();
+    my $child_id = (keys %$children)[0];
+    mutex_wait('verification_ready', $child_id);
```

Comment:
Isn't `wait_for_children` doing exactly same?

Reply:
Nope, didn't work. I assume in cases where both jobs are children, there can be confusion.

Reply:
Let me check code in `os-autoinst`, but I'm surprised as you get id's using `get_children` call. What scares me most is L138 as we rely on the keys order there, looks a bit too hacky. I wonder if there is a better way.

Reply:
So `wait_for_children` just waits till child job is done, so in theory we could use it as we can validate server side with client side disconnected. I've pasted a code below, so what exactly doesn't work if that one is used? I also believe that we don't need `nis_nfs_client_ready` mutex at all, we will just wait till client validation is done and proceed here.

 But why are you saying that both jobs are children? For server side,  is `get_children()` returning 2 elements or just single one?



```perl
sub wait_for_children {
    while (1) {
        my $children = get_children();
        my $n        = 0;
        for my $state (values %$children) {
            next if $state eq 'done' or $state eq 'cancelled';
            $n++;
        }

        bmwqemu::diag("Waiting for $n jobs to finish");
        last unless $n;
        sleep 1;
    }
}
```

Reply:
from my understanding of the locks, it doesn't look like we would need any additional one or modify anything (I might be wrong though):
- Jobs are triggered and client start after server
- Client wait until server creates mutex 'nis_ready'.
- Server continue and client as well and client waits again until server creates locks 'nfs_ready'.
- Server stop waiting for children, just for the client.
- Client continue (here we can add the new module with client validation)
- Client finished (even it could fail retrieving logs)
- Server continue and we could add the server validation.

Doesn't this flow work for some reason?

Reply:
Believe me when I tell you that if it worked, I wouldn't touch it. 
Following,  I have replaced the commended L137-139 with wait_for_children. The test waits forever for 1 job to finish, even though I used mutex_lock on the client and it has moved on to the next module. Exactly the same happened with original code (difference now is that there is only one mutex_create in the server instead of two).
https://openqa.suse.de/tests/4369116

So if wait_for_children is not waiting for a mutex_lock neither the end of the child module, what is it waiting for? The way it is now, it looks like it's waiting for the end of the child test.

Reply:
I will try to add a mutex_unlock at the end of the client module, I don't remember if I already tried that.

Reply:
As per our discussion, we can resolve this one, as `wait_for_children` cannot be used here at all. I've checked that we don't schedule this test anywhere else, so we can safely introduce these changes.